### PR TITLE
Login view

### PR DESCRIPTION
**Pull request by adrian-ponce-wolox**
## Summary ##
* Login view implementation
* Login button disabled if the fields are invalid
* After successful login, user is redirected to dashboard
* Not logged-in users cannot access private urls like `/book/:id` or `/dashboard`. They will redirected to `/login` instead
* Logged-in user cannot access `/login`, they will be redirected to `/dashboard` instead
* The session should persist even when the page is refreshed

screen shot 2017-09-20 at 5 14 36 pm

screen shot 2017-09-20 at 5 16 21 pm

screen shot 2017-09-20 at 5 16 59 pm

### CODE REVIEW

File: w-books/src/App/screens/Login/index.js [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+                  });
+  }
+  validateEmail = () => {
+    return (this.state.email.length > 0 && this.state.email.includes("@"))
+  }
+  validatePassword = () => {
+    if (this.state.password.length >= 8 && this.state.password.length <= 52) {
+      const regLetters = new RegExp("[^a-z|^A-Z]");
+      const regNumbers = new RegExp("[^0-9]");
+      return regLetters.test(this.state.password) && regNumbers.test(this.state.password);
```
Comment (IgnacioAbadie) [inline w-books/src/App/screens/Login/index.js]: extract Regs into a utils file `src/utils/regularExpression.js`

File: w-books/src/App/screens/Login/index.js [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+                  password: this.state.password })
+                  .then(() => {
+                    this.props.history.push('/dashboard')
+                  })
+                  .catch((error) => {
+                    alert (error);
+                  });
+  }
+  validateEmail = () => {
+    return (this.state.email.length > 0 && this.state.email.includes("@"))
```
Comment (IgnacioAbadie) [inline w-books/src/App/screens/Login/index.js]: Reg file

File: w-books/src/App/screens/Login/index.js [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    if (this.state.password.length >= 8 && this.state.password.length <= 52) {
+      const regLetters = new RegExp("[^a-z|^A-Z]");
+      const regNumbers = new RegExp("[^0-9]");
+      return regLetters.test(this.state.password) && regNumbers.test(this.state.password);
+    }
+    return false
+  }
+  render() {
+    const inputValid = this.validateEmail() && this.validatePassword()
+    return (
```
Comment (IgnacioAbadie) [inline w-books/src/App/screens/Login/index.js]: `<Login inputEmail={this.inputEmail} inputPassword={this.inputPassword} onSubmit={this.submit} />`

Comment (IgnacioAbadie) [reply w-books/src/App/screens/Login/index.js]: smart component on `login/index.js` and dumb component in `login/layout.js`

File: w-books/src/config/routes.js [hunk 4]
```diff
@@ -2,17 +2,51 @@ import React from 'react';
...
+    ) : (
+      <Redirect to={{
+        pathname: '/login',
+        state: { from: props.location }
+      }}/>
+    )
+  )}/>
+)
+
+const PublicRoute = ({ component: Component, ...rest }) => (
```
Comment (IgnacioAbadie) [inline w-books/src/config/routes.js]: extract this functions ( `PublicRoute` and `PrivateRoute` ) into a external file with this names each.

**Review by IgnacioAbadie — CHANGES_REQUESTED** [review]

File: w-books/src/App/screens/BookDetail/index.js [hunk 5]
```diff
@@ -16,7 +15,6 @@ function BookDetail({ match }) {
   let book = BookList.find((element) => { return element.id == match.params.bookId})
```
Comment (IgnacioAbadie) [inline w-books/src/App/screens/BookDetail/index.js]: again, move code to `index.js` and ´layout.js`. Do it for all views.

So, do `getSuggestions` and `BookList.find((element) => { return element.id == match.params.bookId})` in `componentWillMount` and store data in state, then pass data to child

File: w-books/src/config/routes.js [hunk 6]
```diff
@@ -41,6 +42,8 @@ const PublicRoute = ({ component: Component, ...rest }) => (
 
 function AppRoutes() {
   return (
+    <div>
+      {isAuthenticated() ? <Header /> : null }
       <Router>
         <div>
         <Redirect to={{pathname: '/login'}}/>
```
Comment (IgnacioAbadie) [inline w-books/src/config/routes.js]: we dont remove this?

File: w-books/src/App/screens/BookDetail/index.js [hunk 7]
```diff
@@ -1,19 +1,14 @@
...
-import Header from '../../components/Header/index.js';
 import Comments from  './components/Comments/index.js';
 import Information from  './components/Information/index.js';
 import Sugestions from  './components/Sugestions/index.js';
 import BookList from '../../../assets/dummy-books.json';
 import CommentList from '../../../assets/dummy-comments.json';
 
 function BookDetail({ match }) {
-  let book = BookList.find((element) => { return element.id == match.params.bookId})
+  let book = BookList.find((element) => { return element.id === match.params.bookId})
```
Comment (IgnacioAbadie) [inline w-books/src/App/screens/BookDetail/index.js]: again, move code to index.js and ´layout.js`. Do it for all views.

So, do getSuggestions and BookList.find((element) => { return element.id == match.params.bookId}) in componentWillMount and store data in state, then pass data to child

**Review by IgnacioAbadie — CHANGES_REQUESTED** [review]

File: w-books/src/App/screens/BookDetail/index.js (hunk 7 above)
Comment (adrian-ponce-wolox) [reply w-books/src/App/screens/BookDetail/index.js]: Done!

File: w-books/src/config/components/PrivateRoute/index.js [hunk 8]
```diff
@@ -0,0 +1,24 @@
+import React from 'react';
```
Comment (IgnacioAbadie) [inline w-books/src/config/components/PrivateRoute/index.js]: Great! just a detail, This arent components, are routes... So i think that this is better.
```
config
   |  routes.js
   |  PrivateRoute.js
   |  PublicRoute.js

```

All as simple files into config folder

File: w-books/src/App/screens/Login/layout.js [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+import wlogo from '../../../assets/Header/wbooks_logo.svg';
+
+function Login({ inputEmail, inputPassword, onSubmit, inputValid }) {
+  return (
+    <div className="Login">
+      <div className="Login-container">
+        <img src={wlogo} className="Login-logo" alt="WBooks" />
+        <h3 className="Login-input-label">Email:</h3>
+        <input className="Login-input" type="text" placeholder="example@wolox.com.ar" onChange={inputEmail}/>
+        <h3 className="Login-input-label">Password:</h3>
```
Comment (IgnacioAbadie) [inline w-books/src/App/screens/Login/layout.js]: We start now to extract all strings into a strings file in same folder that component... so
```
Login
   | index.js
   | layout.js
   | styles.css
   | strings.js
```
where you have centraliced all component!!! string
```
export const password = 'Pasword: ';
export const email = 'Email: '
etc..
```

and then importing this string constants into your components!

Keep calm and make this little refactor in all components

**Review by IgnacioAbadie — CHANGES_REQUESTED** [review]
Its a really great job! :clap: some things more to refine a little more :smile:

File: w-books/src/App/screens/BookDetail/index.js [hunk 10]
```diff
@@ -1,17 +1,24 @@
...
 
-function BookDetailContainer({ match }) {
-  console.log("BookList: " + BookList)
-  console.log("match.params.bookId" + match.params.bookId)
-  let book = BookList.find((element) => { return element.id == match.params.bookId})
-  return <BookDetail book={book}
-                     suggestions={getSuggestions()}
-                     comments={CommentList} />
+class BookDetailContainer extends Component {
+  constructor({match}) {
```
Comment (IgnacioAbadie) [inline w-books/src/App/screens/BookDetail/index.js]: remember you can decalre state like
```
state = { bookId: match.params.bookId, book: null }
```
without use construct

File: w-books/src/App/screens/BookDetail/index.js [hunk 11]
```diff
@@ -1,17 +1,24 @@
...
+    super()
+    this.state = { bookId: match.params.bookId, book: null }
+  }
+  componentWillMount() {
+    let book = BookList.find((element) => { return element.id == this.state.bookId})
```
Comment (IgnacioAbadie) [inline w-books/src/App/screens/BookDetail/index.js]: use const for this

**Review by IgnacioAbadie — APPROVED** [review]
Great, tooooooo better!!
Just check and fix the comments and then lets go to merge 😄

**Pull request merged**